diff --git a/Readme.md b/Readme.md
index e69de29..bd2f531 100644
--- a/Readme.md
+++ b/Readme.md
@@ -0,0 +1,6 @@
+# Crud Xamarin.Forms
+
+Basic CRUD application make in Xamarin.Forms
+
+## References
+- [implement NavigationService](https://www.youtube.com/watch?v=nZxBihE7ByM)
\ No newline at end of file
diff --git a/crud_xamarin/App.xaml.cs b/crud_xamarin/App.xaml.cs
index 9a4e18b..8bba332 100644
--- a/crud_xamarin/App.xaml.cs
+++ b/crud_xamarin/App.xaml.cs
@@ -1,19 +1,58 @@
-﻿using crud_xamarin.Views;
+﻿using crud_xamarin.Models;
+using crud_xamarin.Services;
+using crud_xamarin.ViewModels;
+using crud_xamarin.Views;
+using Microsoft.Extensions.DependencyInjection;
 using System;
 using Xamarin.Forms;
+using Xamarin.Forms.Internals;
 using Xamarin.Forms.Xaml;
 
+
 namespace crud_xamarin
 {
     public partial class App : Application
     {
+        private static IServiceProvider s_serviceProvider;
+
         public App()
         {
             InitializeComponent();
 
-            MainPage = new AppShell();
+            #region Dependency Inyection
+
+            var services = new ServiceCollection();
+
+            // services
+            services.AddSingleton<INavigationService, NavigationService>();
+            services.AddSingleton<IProductService<ProductModel>, MockProductService>();
+            
+            // viewmodels
+            services.AddTransient<ProductsViewModel>();
+            services.AddTransient<ProductDetailViewModel>();
+            services.AddTransient<ProductCreateViewModel>();
+
+            // views
+            services.AddTransient<ProductsView>();
+            services.AddTransient<ProductDetailView>();
+            services.AddTransient<ProductCreateView>();
+
+            services.AddTransient<MainPage>(sp => new MainPage(s_serviceProvider));
+            services.AddSingleton<AppShell>();
+
+            s_serviceProvider = services.BuildServiceProvider();
+            DependencyResolver.ResolveUsing(type => s_serviceProvider.GetService(type));
+
+            #endregion
+
+            //MainPage = new AppShell();
+            //MainPage = s_serviceProvider.GetRequiredService<AppShell>();
+
+            var shell = s_serviceProvider.GetRequiredService<AppShell>();
+            MainPage = shell;
         }
 
+
         protected override void OnStart()
         {
         }
diff --git a/crud_xamarin/AppShell.xaml.cs b/crud_xamarin/AppShell.xaml.cs
index 5b8e5e8..9b9bd91 100644
--- a/crud_xamarin/AppShell.xaml.cs
+++ b/crud_xamarin/AppShell.xaml.cs
@@ -14,11 +14,23 @@ namespace crud_xamarin
     [XamlCompilation(XamlCompilationOptions.Compile)]
     public partial class AppShell : Shell
     {
-        public AppShell()
+        private readonly IServiceProvider _serviceProvider;
+
+        //public AppShell()
+        //{
+        //    InitializeComponent();
+
+        //    //RegisterServices();
+        //    RegisterRoutes();
+        //}
+
+        public AppShell(IServiceProvider serviceProvider)
         {
             InitializeComponent();
 
-            RegisterServices();
+            _serviceProvider = serviceProvider;
+
+            //RegisterServices();
             RegisterRoutes();
         }
 
@@ -29,9 +41,11 @@ namespace crud_xamarin
             Routing.RegisterRoute(nameof(ProductCreateView), typeof(ProductCreateView));
         }
 
+        [Obsolete]
         private void RegisterServices()
         {
             DependencyService.Register<MockProductService>();
+            DependencyService.Register<INavigationService, NavigationService>();
         }
     }
 }
\ No newline at end of file
diff --git a/crud_xamarin/MainPage.xaml.cs b/crud_xamarin/MainPage.xaml.cs
index f49443f..82eff0d 100644
--- a/crud_xamarin/MainPage.xaml.cs
+++ b/crud_xamarin/MainPage.xaml.cs
@@ -11,9 +11,18 @@ namespace crud_xamarin
 {
     public partial class MainPage : ContentPage
     {
-        public MainPage()
+        private readonly IServiceProvider _serviceProvider;
+
+        //public MainPage()
+        //{
+        //    InitializeComponent();
+        //}
+
+        public MainPage(IServiceProvider serviceProvider)
         {
             InitializeComponent();
+
+            _serviceProvider = serviceProvider;
         }
 
         private async void btnGoToProducts_Clicked(object sender, EventArgs e)
diff --git a/crud_xamarin/Services/INavigationService.cs b/crud_xamarin/Services/INavigationService.cs
new file mode 100644
index 0000000..215e60c
--- /dev/null
+++ b/crud_xamarin/Services/INavigationService.cs
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace crud_xamarin.Services
+{
+    public interface INavigationService
+    {
+        Task GoToAsync(string route, object parameter = null);
+        Task GoBackAsync();
+        List<string> GetHistory();
+    }
+}
diff --git a/crud_xamarin/Services/NavigationService.cs b/crud_xamarin/Services/NavigationService.cs
new file mode 100644
index 0000000..2fc2667
--- /dev/null
+++ b/crud_xamarin/Services/NavigationService.cs
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace crud_xamarin.Services
+{
+    public class NavigationService : INavigationService
+    {
+        private List<string> _history;
+        private readonly Shell _shell;
+        
+        public NavigationService(Shell shell)
+        {
+            _shell = shell;
+            _history = new List<string>();
+        }
+
+        public async Task GoBackAsync()
+        {
+            await _shell.GoToAsync("..");
+            _history.RemoveAt(_history.Count);
+        }
+
+        public async Task GoToAsync(string route, object parameter = null)
+        {
+            await _shell.GoToAsync(route);
+            _history.Add(route);
+        }
+
+        public List<string> GetHistory()
+        {
+            return _history;
+        }
+    }
+}
diff --git a/crud_xamarin/ViewModels/BaseViewModel.cs b/crud_xamarin/ViewModels/BaseViewModel.cs
index d7c78a9..755711f 100644
--- a/crud_xamarin/ViewModels/BaseViewModel.cs
+++ b/crud_xamarin/ViewModels/BaseViewModel.cs
@@ -6,7 +6,7 @@ using System.Text;
 
 namespace crud_xamarin.ViewModels
 {
-    class BaseViewModel : INotifyPropertyChanged
+    public class BaseViewModel : INotifyPropertyChanged
     {
         public event PropertyChangedEventHandler PropertyChanged;
 
diff --git a/crud_xamarin/ViewModels/ProductCreateViewModel.cs b/crud_xamarin/ViewModels/ProductCreateViewModel.cs
index 3b21e3c..583deac 100644
--- a/crud_xamarin/ViewModels/ProductCreateViewModel.cs
+++ b/crud_xamarin/ViewModels/ProductCreateViewModel.cs
@@ -9,7 +9,7 @@ using Xamarin.Forms;
 
 namespace crud_xamarin.ViewModels
 {
-    class ProductCreateViewModel: BaseViewModel
+    public class ProductCreateViewModel: BaseViewModel
     {
         private ProductModel _product = new ProductModel();
         public ProductModel Product
diff --git a/crud_xamarin/ViewModels/ProductDetailViewModel.cs b/crud_xamarin/ViewModels/ProductDetailViewModel.cs
index 20f4ff4..5df49b2 100644
--- a/crud_xamarin/ViewModels/ProductDetailViewModel.cs
+++ b/crud_xamarin/ViewModels/ProductDetailViewModel.cs
@@ -9,7 +9,7 @@ using Xamarin.Forms;
 namespace crud_xamarin.ViewModels
 {
     [QueryProperty("Id", nameof(Id))]
-    class ProductDetailViewModel : BaseViewModel
+    public class ProductDetailViewModel : BaseViewModel
     {
         private int _id;
         public int Id
diff --git a/crud_xamarin/ViewModels/ProductsViewModel.cs b/crud_xamarin/ViewModels/ProductsViewModel.cs
index ebda7c4..9b8fd34 100644
--- a/crud_xamarin/ViewModels/ProductsViewModel.cs
+++ b/crud_xamarin/ViewModels/ProductsViewModel.cs
@@ -11,7 +11,7 @@ using Xamarin.Forms;
 
 namespace crud_xamarin.ViewModels
 {
-    class ProductsViewModel : BaseViewModel
+    public class ProductsViewModel : BaseViewModel
     {
         private ObservableCollection<ProductModel> _products;
         public ObservableCollection<ProductModel> Products
diff --git a/crud_xamarin/Views/ProductCreateView.xaml b/crud_xamarin/Views/ProductCreateView.xaml
index 93e2654..2e4621c 100644
--- a/crud_xamarin/Views/ProductCreateView.xaml
+++ b/crud_xamarin/Views/ProductCreateView.xaml
@@ -4,9 +4,9 @@
              xmlns:viewmodels="clr-namespace:crud_xamarin.ViewModels"
              x:Class="crud_xamarin.Views.ProductCreateView">
 
-    <ContentPage.BindingContext>
+    <!--<ContentPage.BindingContext>
         <viewmodels:ProductCreateViewModel/>
-    </ContentPage.BindingContext>
+    </ContentPage.BindingContext>-->
 
     <ScrollView>
         <StackLayout Padding="10">
diff --git a/crud_xamarin/Views/ProductCreateView.xaml.cs b/crud_xamarin/Views/ProductCreateView.xaml.cs
index 21e5bec..a64a0cc 100644
--- a/crud_xamarin/Views/ProductCreateView.xaml.cs
+++ b/crud_xamarin/Views/ProductCreateView.xaml.cs
@@ -16,10 +16,16 @@ namespace crud_xamarin.Views
         private readonly ProductCreateViewModel _viewModel;
 
         public ProductCreateView()
+        {
+            InitializeComponent();
+        }
+
+        public ProductCreateView(ProductCreateViewModel viewModel)
         {
             InitializeComponent();
 
-            _viewModel = (ProductCreateViewModel)BindingContext;
+            BindingContext = viewModel;
+            _viewModel = viewModel;
         }
     }
 }
\ No newline at end of file
diff --git a/crud_xamarin/Views/ProductDetailView.xaml b/crud_xamarin/Views/ProductDetailView.xaml
index 0362461..b78a9e6 100644
--- a/crud_xamarin/Views/ProductDetailView.xaml
+++ b/crud_xamarin/Views/ProductDetailView.xaml
@@ -5,9 +5,9 @@
              x:Class="crud_xamarin.Views.ProductDetailView"
              x:DataType="viewmodels:ProductDetailViewModel">
 
-    <ContentPage.BindingContext>
+    <!--<ContentPage.BindingContext>
         <viewmodels:ProductDetailViewModel/>
-    </ContentPage.BindingContext>
+    </ContentPage.BindingContext>-->
     
     <StackLayout Padding="10">
         <Label Text="{Binding Product.Id}"
diff --git a/crud_xamarin/Views/ProductDetailView.xaml.cs b/crud_xamarin/Views/ProductDetailView.xaml.cs
index 7b4d3b9..1d767ea 100644
--- a/crud_xamarin/Views/ProductDetailView.xaml.cs
+++ b/crud_xamarin/Views/ProductDetailView.xaml.cs
@@ -14,12 +14,18 @@ namespace crud_xamarin.Views
     public partial class ProductDetailView : ContentPage
     {
         private readonly ProductDetailViewModel _viewModel;
-
+        
         public ProductDetailView()
+        {
+            InitializeComponent();
+        }
+
+        public ProductDetailView(ProductDetailViewModel viewModel)
         {
             InitializeComponent();
 
-            _viewModel = (ProductDetailViewModel)BindingContext;
+            BindingContext = viewModel;
+            _viewModel = viewModel;
         }
 
         protected override async void OnAppearing()
diff --git a/crud_xamarin/Views/ProductsView.xaml b/crud_xamarin/Views/ProductsView.xaml
index a60eb1e..95e4ac5 100644
--- a/crud_xamarin/Views/ProductsView.xaml
+++ b/crud_xamarin/Views/ProductsView.xaml
@@ -5,9 +5,9 @@
              xmlns:models="clr-namespace:crud_xamarin.Models"
              x:Class="crud_xamarin.Views.ProductsView">
 
-    <ContentPage.BindingContext>
+    <!--<ContentPage.BindingContext>
         <viewmodels:ProductsViewModel/>
-    </ContentPage.BindingContext>
+    </ContentPage.BindingContext>-->
 
     <ContentPage.Resources>
         <Style TargetType="Label" x:Key="LabelCellStyle" BasedOn="{StaticResource LabelDefaultStyle}">
diff --git a/crud_xamarin/Views/ProductsView.xaml.cs b/crud_xamarin/Views/ProductsView.xaml.cs
index 10c1846..37395ae 100644
--- a/crud_xamarin/Views/ProductsView.xaml.cs
+++ b/crud_xamarin/Views/ProductsView.xaml.cs
@@ -16,10 +16,16 @@ namespace crud_xamarin.Views
         private readonly ProductsViewModel _viewModel;
 
         public ProductsView()
+        {
+            InitializeComponent();
+        }
+
+        public ProductsView(ProductsViewModel viewModel)
         {
             InitializeComponent();
 
-            _viewModel = (ProductsViewModel)BindingContext;
+            BindingContext = viewModel;
+            _viewModel = viewModel;
         }
 
         protected override async void OnAppearing()
diff --git a/crud_xamarin/crud_xamarin.csproj b/crud_xamarin/crud_xamarin.csproj
index df383d1..7edd869 100644
--- a/crud_xamarin/crud_xamarin.csproj
+++ b/crud_xamarin/crud_xamarin.csproj
@@ -11,6 +11,8 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Autofac" Version="8.4.0" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.1" />
     <PackageReference Include="Xamarin.Forms" Version="5.0.0.2012" />
     <PackageReference Include="Xamarin.Essentials" Version="1.6.1" />
   </ItemGroup>
